src/main/java/org/jenkinsci/plugins/workflow/multibranch/DurabilityHintBranchProperty.java

```diff
@@ -39,44 +45,63 @@ public DurabilityHintBranchProperty(@Nonnull FlowDurabilityHint hint) {
-        if (!clazz.isAssignableFrom(WorkflowJob.class)) {
-            return null;
-        }
-        return new JobDecorator<P, B>() {
-            @NonNull
-            @Override
-            public List<JobProperty<? super P>> jobProperties(
-                    @NonNull List<JobProperty<? super P>> jobProperties) {
-                List<JobProperty<? super P>> result = asArrayList(jobProperties);
-                for (Iterator<JobProperty<? super P>> iterator = result.iterator(); iterator.hasNext(); ) {
-                    JobProperty<? super P> p = iterator.next();
-                    if (p instanceof DurabilityHintJobProperty) {
-                        iterator.remove();
-                    }
-                }
-                if (getHint() != null) {
-                    result.add((JobProperty)(new DurabilityHintJobProperty(getHint())));
-                }
-                return result;
-            }
-        };
+        return null;
     }
 
     @Extension
-    public static class DescriptorImpl extends BranchPropertyDescriptor {
+    public static class DescriptorImpl extends BranchPropertyDescriptor implements DurabilityHintProvider {
         @Override
         public String getDisplayName() {
-            return "Pipeline speed/durability override";
+            return "Pipeline branch speed/durability override";
         }
 
         public FlowDurabilityHint[] getDurabilityHintValues() {
             return FlowDurabilityHint.values();
         }
 
-        public static FlowDurabilityHint getDefaultHint() {
+        public static FlowDurabilityHint getDefaultDurabilityHint() {
             return GlobalDefaultFlowDurabilityLevel.getDefaultDurabilityHint();
         }
+
+        @Override
+        /** Lower ordinal than {@link org.jenkinsci.plugins.workflow.job.properties.DurabilityHintJobProperty} so those can override. */
+        public int ordinal() {
+            return 200;
+        }
+
+        @CheckForNull
+        @Override
+        /**
+         * Dynamically fetch the property with each build, because the {@link BranchPropertyStrategy} does not re-evaluate,
+         * resulting in {@see <a href="https://issues.jenkins-ci.org/browse/JENKINS-48826">JENKINS-48826</a>}. */
+        public FlowDurabilityHint suggestFor(@Nonnull Item x) {
+            // BranchJobProperty *should* be present if it's a child of a MultiBranchProject but we double-check for safety
+            if (x instanceof WorkflowJob && x.getParent() instanceof MultiBranchProject && ((WorkflowJob)x).getProperty(BranchJobProperty.class) != null) {
+                MultiBranchProject mp  = (MultiBranchProject)(x.getParent());
+                WorkflowJob job = (WorkflowJob)x;
+                BranchJobProperty bjp = job.getProperty(BranchJobProperty.class);
+
+                String sourceId = bjp.getBranch().getSourceId();
+                if (sourceId != null) {
```

Comment:
Huh, surprised that this is potentially nullable. I think in practice it is not, but better to be defensive until the annotations are cleaned up upstream.